### PR TITLE
 Allow reserved keywords as identifiers

### PR DESCRIPTION
**Pull request by davidytag**
Addresses https://github.com/google/closure-compiler/issues/2852.

Currently, the compiler will fail to interpret JavaScript future reserved words when they're used as function names and parameters. This PR fixes that issue by allowing the parser to determine when it's appropriate to use reserved words as identifiers.

### CODE REVIEW

**Comment by brad4d** [comment]
@davidytag 
Could you address the review comments?

**Comment by davidytag** [comment]
@brad4d 
I've resolved the merge conflicts - is there anything else you'd like me to review?

**Comment by brad4d** [comment]
@davidytag 
sorry, yes I had made review comments
I suspect you didn't see them because I failed to hit the "finish review" button.

File: src/com/google/javascript/jscomp/parsing/parser/Parser.java [hunk 1]
```diff
@@ -285,6 +285,13 @@ private boolean isKeyword(String value) {
         && (!Keywords.isTypeScriptSpecificKeyword(value) || config.parseTypeSyntax);
   }
 
+  /**
+   * Returns true if strict mode is enabled and the next token is a strict keyword.
```
[the remark this answers is not in the document]
Comment (davidytag) [reply src/com/google/javascript/jscomp/parsing/parser/Parser.java]: Ah yes, I meant to say "if strict mode is disabled". Making both changes you're suggesting.

File: src/com/google/javascript/jscomp/parsing/parser/Parser.java [hunk 2]
```diff
@@ -2833,11 +2840,15 @@ private ParseTree parse(Expression expressionIn) {
...
+    if (peekReservedWordAsId()) {
+      return parseIdentifierExpression();
+    }
+
     return parseAssignment(Expression.NORMAL);
   }
 
   private boolean peekAssignmentExpression() {
-    return peekExpression();
+    return peekExpression() || peekReservedWordAsId();
```
[the remark this answers is not in the document]
Comment (davidytag) [reply src/com/google/javascript/jscomp/parsing/parser/Parser.java]: That sounds right. I agree with your previous comments above... it doesn't really make sense to process a keyword that can be treated as an identifier as an assignment expression.

**Comment by davidytag** [comment]
@brad4d Has this been fixed? I pulled from master recently and I'm getting a correct result when strict mode is off.

**Comment by brad4d** [comment]
@davidytag I'm concerned we may be out of sync here.
I believe I'm waiting for you to finish addressing review comments.
Do you think otherwise?
Thanks.

**Comment by brad4d** [comment]
Created internal issue b/122075547

**Pull request closed without merging**